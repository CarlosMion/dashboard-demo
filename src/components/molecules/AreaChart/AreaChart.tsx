import { BalanceHistory } from "@/types";
import React from "react";
import {
  AreaChart as RechartsAreaChart,
  Area,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  ResponsiveContainer,
} from "recharts";
import { Container } from "./styled";
import { useTheme } from "@mui/material";
import { useTranslations } from "next-intl";

interface AreaChartProps {
  data: BalanceHistory[];
}

export default function AreaChart({ data }: AreaChartProps) {
  const theme = useTheme();
  const t = useTranslations("card");
  return (
    <Container>
      <ResponsiveContainer width="100%" height="100%">
        <RechartsAreaChart
          data={data}
          margin={{
            top: 20,
            right: 30,
            left: 20,
            bottom: 5,
          }}
        >
          <CartesianGrid
            horizontal={true}
            vertical={true}
            strokeDasharray="3 3"
          />
          <XAxis dataKey="date" axisLine={false} tickLine={true} />
          <YAxis
            axisLine={false}
            tickLine={true}
            tickFormatter={(value) => value.toLocaleString()}
          />
          <Tooltip
            formatter={(value) => [`$${value.toLocaleString()}`, t("balance")]}
            labelStyle={{ color: theme.palette.text.secondary }}
          />
          <Area
            type="monotone"
            dataKey="balance"
            stroke={theme.palette.common.blue}
            fill="url(#colorGradient)"
            strokeWidth={2}
          />
          <defs>
            <linearGradient id="colorGradient" x1="0" y1="0" x2="0" y2="1">
              <stop
                offset="5%"
                stopColor={theme.palette.common.blue}
                stopOpacity={0.8}
              />
              <stop
                offset="95%"
                stopColor={theme.palette.common.blue}
                stopOpacity={0.1}
              />
            </linearGradient>
          </defs>
        </RechartsAreaChart>
      </ResponsiveContainer>
    </Container>
  );
}
