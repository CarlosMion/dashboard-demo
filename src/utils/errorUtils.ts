export type ServerError = {
  error: string;
  statusCode?: number;
};

export const isServerErrorWithStatusCode = (
  result: object
): result is ServerError => {
  return "error" in result && "statusCode" in result;
};

export const isServerError = (result: unknown): result is ServerError => {
  return typeof result === "object" && result !== null && "error" in result;
};

type ErrorWithMessage = {
  message: string;
};

export const isErrorWithMessage = (
  error: unknown
): error is ErrorWithMessage => {
  return (
    typeof error === "object" &&
    error !== null &&
    "message" in error &&
    typeof (error as Record<string, unknown>).message === "string"
  );
};

export const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage => {
  if (isErrorWithMessage(maybeError)) return maybeError;

  try {
    return new Error(JSON.stringify(maybeError));
  } catch {
    // fallback in case there's an error stringifying the maybeError
    // like with circular references for example.
    return new Error(String(maybeError));
  }
};

export const getErrorMessage = (error: unknown): string => {
  return toErrorWithMessage(error).message;
};

export class ApiError extends Error {
  statusCode: number;

  constructor(message: string, statusCode: number) {
    super(message);
    this.name = "ApiError";
    this.statusCode = statusCode || 500;
  }
}
