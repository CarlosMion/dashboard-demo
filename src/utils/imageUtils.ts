export const dataURItoBlob = (dataURI: string, fileType: string): Blob => {
  const byteString = atob(dataURI.split(',')[1]);
  const arrayBuffer = new ArrayBuffer(byteString.length);
  const intArray = new Uint8Array(arrayBuffer);

  for (let i = 0; i < byteString.length; i++) {
    intArray[i] = byteString.charCodeAt(i);
  }

  return new Blob([arrayBuffer], { type: fileType });
};

export const dataURItoFile = (
  dataURI: string,
  fileName: string,
  fileType: string
): File => {
  const blob = dataURItoBlob(dataURI, fileType);
  return new File([blob], fileName, { type: blob.type });
};

export const blobToFile = (blob: Blob, fileName: string): File => {
  return new File([blob], fileName, { type: blob.type });
};

export const getImageDimensions = (
  src: string
): Promise<{ width: number; height: number } | null> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = src;

    img.onload = () =>
      resolve(
        img.width && img.height
          ? { width: img.width, height: img.height }
          : null
      );
    img.onerror = (error) => reject(error);
  });
};

export const isImageDimensionWithinThreshold = (
  actual: number,
  ideal: number,
  threshold: number
): boolean => {
  const isWithinThreshold = actual >= ideal * (1 - threshold);
  return isWithinThreshold;
};
