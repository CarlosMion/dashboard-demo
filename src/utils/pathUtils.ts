export function toCamelCase(str?: string): string | undefined {
  return str?.replace(/-([a-z])/g, (match, letter) => letter.toUpperCase());
}
