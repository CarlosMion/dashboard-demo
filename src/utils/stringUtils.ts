export function maskCardNumber(cardNumber: number): string {
  const cardNumberStr = cardNumber.toString();
  return cardNumberStr.replace(/(\d{4})\d{8}(\d{4})/, "$1 **** **** $2");
}

export function formatMoney(value: number): string {
  return value.toLocaleString("en-US", { style: "currency", currency: "USD" });
}

export function formatValidThru(value: string): string {
  return new Date(value).toLocaleDateString("en-US", {
    month: "2-digit",
    year: "2-digit",
    timeZone: "UTC",
  });
}

export function formatTransactionDate(value: string): string {
  return new Date(value).toLocaleDateString("en-US", {
    day: "2-digit",
    month: "long",
    year: "numeric",
    timeZone: "UTC",
  });
}
