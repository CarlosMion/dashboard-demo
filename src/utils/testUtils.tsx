import { render } from "@testing-library/react";
import { ReactNode } from "react";
import AppProviders from "@/providers/AppProviders/AppProviders";

import messages from "@/i18n/messages/en.json";

const customRender = (children: ReactNode) =>
  render(
    <AppProviders
      locale={"en"}
      messages={messages}
      userAgent="Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    >
      {children}
    </AppProviders>
  );

export * from "@testing-library/react";

export { customRender as render };
