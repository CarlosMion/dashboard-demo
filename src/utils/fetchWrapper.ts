import { isServerError } from "./errorUtils";

const fetcher = async <T>(
  info: RequestInfo | URL,
  init?: RequestInit
): Promise<T> => {
  let result;
  try {
    const injectedHeaders: Record<string, string> = {};
    if (typeof window === "undefined") {
      const headers = await import("next/headers");
      (await headers.headers()).forEach((value, key) => {
        injectedHeaders[key] = value;
      });
    }

    const buildHeaders: Record<string, string> = {};
    let url = info;
    if (typeof info === "string") {
      url = `${process.env.NEXT_PUBLIC_WEBAPP_API_URL}/${info}`;
    }
    const reqHeaders = {
      "Content-Type": "application/json",
      ...init?.headers,
      ...injectedHeaders,
      ...buildHeaders,
    };
    result = await fetch(url, {
      ...init,
      headers: reqHeaders,
    });
    if (!result.ok) {
      throw new Error("Failed to fetch");
    }
    let parsedResult;
    try {
      parsedResult = await result.json();
    } catch (error) {
      console.log("Failed to parse response, probably not JSON", result, error);
      return result as T;
    }
    if (isServerError(parsedResult)) {
      throw new Error(parsedResult.error);
    }
    return parsedResult as T;
  } catch (error) {
    throw error;
  }
};

export default fetcher;
