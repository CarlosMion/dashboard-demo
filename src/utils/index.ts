import { isEdge } from "./userAgentUtils";
import { maskCardNumber } from "./stringUtils";
import { toCamelCase } from "./pathUtils";
import fetcher from "./fetchWrapper";
import {
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
} from "./errorUtils";
import {
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
} from "./imageUtils";
import {
  createImage,
  getRadianAngle,
  rotateSize,
  getCroppedImg,
} from "./cropImageUtils";
export {
  fetcher,
  isEdge,
  toCamelCase,
  isServerErrorWithStatusCode,
  ApiError,
  getErrorMessage,
  isErrorWithMessage,
  isServerError,
  toErrorWithMessage,
  blobToFile,
  dataURItoBlob,
  dataURItoFile,
  getImageDimensions,
  isImageDimensionWithinThreshold,
  getCroppedImg,
  rotateSize,
  getRadianAngle,
  createImage,
  maskCardNumber,
};
