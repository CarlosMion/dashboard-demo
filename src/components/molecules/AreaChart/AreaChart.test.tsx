import { render } from "@/utils/testUtils";
import AreaChart from "./AreaChart";
import { PropsWithChildren } from "react";

const originalError = console.error;
beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  console.error = (...args: any[]) => {
    // if (args[0].includes("Warning: Invalid DOM property")) return;
    if (args[0].includes("incorrect casing")) return;
    originalError.call(console, ...args);
  };
});

afterAll(() => {
  console.error = originalError;
});

const mockHtmlElements = () => {
  const originalCreateElement = document.createElement.bind(document);
  jest.spyOn(document, "createElement").mockImplementation((tagName) => {
    if (
      tagName === "linearGradient" ||
      tagName === "stop" ||
      tagName === "defs"
    ) {
      return originalCreateElement("span");
    }
    return originalCreateElement(tagName);
  });
};

jest.mock("recharts", () => ({
  ResponsiveContainer: ({ children }: PropsWithChildren) => (
    <div>{children}</div>
  ),
  AreaChart: ({ children }: PropsWithChildren) => <div>{children}</div>,
  Area: () => <div>Area</div>,
  XAxis: () => <div>XAxis</div>,
  YAxis: () => <div>YAxis</div>,
  CartesianGrid: () => <div>CartesianGrid</div>,
  Tooltip: () => <div>Tooltip</div>,
}));

describe("AreaChart", () => {
  beforeAll(() => {
    mockHtmlElements();
  });
  const mockData = [
    { date: "2024-01", balance: 1000 },
    { date: "2024-02", balance: 1500 },
  ];

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<AreaChart data={mockData} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders all chart components", () => {
    const { getByText } = render(<AreaChart data={mockData} />);
    expect(getByText("Area")).toBeInTheDocument();
    expect(getByText("XAxis")).toBeInTheDocument();
    expect(getByText("YAxis")).toBeInTheDocument();
    expect(getByText("CartesianGrid")).toBeInTheDocument();
    expect(getByText("Tooltip")).toBeInTheDocument();
  });
});
