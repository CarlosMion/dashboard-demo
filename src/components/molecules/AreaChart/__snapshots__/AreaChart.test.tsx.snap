// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AreaChart renders correctly and matches snapshot 1`] = `
<div
  class="MuiBox-root css-1h8flee"
>
  <div>
    <div>
      <div>
        CartesianGrid
      </div>
      <div>
        XAxis
      </div>
      <div>
        YAxis
      </div>
      <div>
        Tooltip
      </div>
      <div>
        Area
      </div>
      <span>
        <span
          id="colorGradient"
          x1="0"
          x2="0"
          y1="0"
          y2="1"
        >
          <span
            offset="5%"
            stop-color="#2D60FF"
            stop-opacity="0.8"
          />
          <span
            offset="95%"
            stop-color="#2D60FF"
            stop-opacity="0.1"
          />
        </span>
      </span>
    </div>
  </div>
</div>
`;
